Use NPM SimpleSchema rather than Meteor

aldeed opened the pull request:
**BREAKING CHANGES**

This PR updates the `aldeed:simple-schema` Meteor package dependency to instead depend on the `simpl-schema` NPM package, which is the newest release of the same library. As part of this change, there are several breaking changes and other gotchas to be aware of.

- **IMPORTANT!** The NPM package does not play nice with the previous Meteor package. After updating to this Reaction release, run the app one time, and then look at the `.meteor/versions` file. Make sure that `aldeed:simple-schema` is not listed. If it is there, that is because you depend on another Meteor package that depends on `aldeed:simple-schema`. You will have to update or remove any such packages (with `meteor remove` / `meteor add`) until `aldeed:simple-schema` disappears from your `.meteor/versions` file.
- Search your app for any `import { SimpleSchema } from "meteor/aldeed:simple-schema"` lines that you have added in your custom code, and replace them with `import SimpleSchema from "simpl-schema"`
- Be aware that the package name does not have the "e" on "simpl". (There is a different NPM package called `simple-schema` with the "e", and that is NOT the one you want.)
- If you have your own custom schemas, refer to the SimpleSchema changelog to update them for the breaking changes: https://github.com/aldeed/meteor-simple-schema/blob/master/CHANGELOG.md#200
- If you use `attachSchema` in your code, be aware that passing an array to `attachSchema` will no longer work. You should first merge all the schemas and then pass the combined schema to `attachSchema`

zenweasel commented:
@aldeed Hey Eric, I think I speak for everyone when I say we are super-excited to get this work done and to have you doing it. Just starting to review this code now and my first question is that I am unclear on why some of these changes need to be made to support node-simple-schema. I wonder if you could put in some review comments about why some of these changes were made it would make it easier for me to understand what's going on. (this is one area where GH code review is really lacking)

I wonder if maybe we could schedule a time to talk to me and/or the team about this and help us understand how node-simple-schema differs from meteor-simple-schema. It's a pretty core part of our project and everyone on the team needs to have a really good understanding of how it works.

aldeed commented:
@zenweasel There might be some changes still in here from my trying to debug issues, which I will remove before finalizing. Still need to skim through it all a final time myself. I also fixed some over-reactive autoruns while trying to figure out a bug caused by the updates, but I'll move those changes into a separate PR since they're technically unrelated.

I'll add some details about what things were updated as I do another read-through of the diff. Possibly tomorrow or this weekend.

aldeed commented:
@aaronjudd Sorry I had more changes I hadn't pushed. I pushed them now, and should include fixing the error you saw. There are still other more hidden issues I'm investigating, but I think it's getting close.

aldeed commented on imports/plugins/included/payments-stripe/server/methods/stripeapi-methods-capture.app-test.js:
```diff
@@ -149,13 +149,9 @@ describe("stripe/payment/capture", function () {
-      captureError = error;
-      expect(captureError.message).to.equal("Match error: Match error: Transaction id is required");
-      expect(captureResult).to.be.undefined;
+      expect(error.message).to.equal("[Transaction ID is required]");
```
They are no longer "Match error"s since they don't go through the `check` package

aldeed commented on imports/plugins/included/product-detail-simple/client/containers/publish.js:
```diff
@@ -17,12 +17,15 @@ class ProductPublishContainer extends Component {
 
   handleVisibilityChange = (event, isProductVisible) => {
+    const { product } = this.props;
+    if (!product) return;
```
Was seeing error related to undefined product, so I "fixed"

aldeed commented on imports/plugins/included/product-detail-simple/client/containers/publish.js:
```diff
@@ -39,13 +42,16 @@ class ProductPublishContainer extends Component {
-    if (result && result.status === "success" && this.props.product) {
-      const productDocument = result.previousDocuments.find((product) => this.props.product._id === product._id);
+    const { product } = this.props;
+    if (!product) return;
```
Was seeing error related to undefined product, so I "fixed"

aldeed commented on imports/plugins/included/sms/server/methods/sms.js:
```diff
@@ -49,6 +49,7 @@ Meteor.methods({
     check(shopId, String);
 
     const user = Accounts.findOne({ _id: userId });
+    if (!user) return;
```
This was causing an error sometimes when running tests

aldeed commented on lib/collections/schemas/accounts.js:
```diff
@@ -131,93 +136,87 @@ registerSchema("Email", Email);
-    }
-  },
-  updatedAt: {
+    autoValue: createdAtAutoValue
```
Tried to update all the createdAt/updatedAt to use the same functions, unless there seemed to be a good reason for it to be done differently.

aldeed commented on lib/collections/schemas/orders.js:
```diff
@@ -156,34 +161,46 @@ registerSchema("OrderTransaction", OrderTransaction);
 registerSchema("Order", Order);
+
+export const OrderDocument = Cart.clone().extend(Order).extend(OrderItem);
+registerSchema("OrderDocument", OrderDocument);
```
This was formerly handled by passing an array of schemas to `attachSchema` but that no longer works. So now I'm registering it as its own schema and then attaching that.

aldeed commented on server/api/core/core.js:
```diff
@@ -975,53 +978,5 @@ export default {
-   *                  that have been attached to the collection or false if
-   *                  the collection or schema could not be found
-   */
-  collectionSchema(collection, selector) {
```
Not used anymore

zenweasel replied:
Can we provide any sort of deprecation path there rather than just removing this in case application developers are using it?

aldeed commented on server/api/core/import.js:
```diff
@@ -6,7 +6,6 @@ import { MongoInternals } from "meteor/mongo";
 import * as Collections from "/lib/collections";
 import Hooks from "../hooks";
 import { Logger } from "../logger";
-import doRightJoinNoIntersection from "./rightJoin";
```
Deleted this file since it's no longer used here or anywhere

aldeed commented on server/api/core/import.js:
```diff
@@ -194,12 +192,7 @@ Import.context = function (collection, selector) {
-  }
-  this._contexts[name] = collection.simpleSchema(selector).newContext();
-  return this._contexts[name];
+  return collection.simpleSchema(selector).namedContext(name);
```
SS already has a `namedContext` function that keeps a reference in this same way, so it seemed better to just use that

aldeed commented on server/api/core/import.js:
```diff
@@ -410,35 +402,26 @@ Import.object = function (collection, key, object) {
-    });
-  }
+  this.buffer(collection).find(key).upsert().update(cleanedModifier);
+
```
I'm doing all of the above in a different way, but I'm pretty sure it has the same end effect

aldeed commented on server/imports/fixtures/orders.js:
```diff
@@ -130,17 +130,20 @@ export default function () {
     requiresShipping: true,
     shipping: [{
       shopId: getShopId(),
+      address: getAddress({ isShippingDefault: true }),
```
These missing props were causing test failures. I'm pretty sure they were always improperly missing, but for some reason the old SS package didn't catch the errors I guess.

aldeed commented on server/methods/catalog.js:
```diff
@@ -154,9 +154,15 @@ function createHandle(productHandle, productId) {
+    _id: {
+      $nin: [productId]
+    }
+  }).count();
```
This fixed an infinite loop I was seeing, though I don't recall how to reproduce it. It should be obvious reading the code that the $nin should have been there.

aldeed commented on server/methods/core/orders.app-test.js:
```diff
@@ -527,8 +527,8 @@ describe("orders test", function () {
         expect(orderPaymentMethod.status).to.equal("completed");
+        done();
       });
-      return done();
```
`done` was called too early with these on the outside

aldeed commented on server/methods/core/orders.js:
```diff
@@ -974,51 +974,57 @@ export const methods = {
+        result = Meteor.call(`${processor}/payment/capture`, paymentMethod);
+      } catch (e) {
+        error = e;
+      }
```
The only change here is that the Meteor.call of `${processor}/payment/capture` is now synchronous (has no callback). Before this change, I had tests failing because that call would still be executing after the test had already finished and removed the sandbox data. Doing it synchronous seems like it would be better in real world, too, but on the off chance that this was intentionally async, there might be a different solution by rewriting how the tests work.

aldeed commented on server/startup/init.js:
```diff
@@ -1,6 +1,18 @@
+  ddpError.error = "validation-error";
+  ddpError.details = error.details;
+  return ddpError;
+});
```
Not sure if this is the proper file to put this in. Let me know if you want it moved.

aldeed commented:
@zenweasel @aaronjudd I think this is pretty close. Finally figured out all the pesky bugs and got tests passing (on my machine at least). This was much more difficult than I thought, mainly because all of your crazy schemas illuminated several edge case bugs in SimpleSchema. But the good news is that I feel quite confident in npm SimpleSchema now.

I ran through some screens, but I'm not an expert on the app workflows, so it would be helpful to have some manual testing through common flows and with common plugins.

zenweasel commented:
Seeing this error on startup:

```
23:07:51.982Z  WARN Reaction: Error while importing to Packages: Cannot update 'settings' and 'settings' at the same time
```

zenweasel commented:
See this error when I add a product and set the price:

```
Match error: Failed Match.OneOf, Match.Maybe or Match.Optional validation [server-error]
```

zenweasel commented on imports/plugins/core/checkout/client/methods/cart.js:
```diff
@@ -11,7 +11,7 @@ Meteor.methods({
   // Under consideration for deprecation and migrating other payment Packages
   // to payments-stripe style methods
   "cart/submitPayment": function (paymentMethod) {
-    check(paymentMethod, Reaction.Schemas.PaymentMethod);
```
I think this is actually a pretty important check here that shouldn't just be switched to `Object`. If the `paymentMethod` doesn't have specific values it's going to fail hard upstream and we should catch it here.

aldeed replied:
@zenweasel It might be possible to use the schema, but the problem I was having with using the schema in some places like this is that the payment method being passed is somewhat different from what ultimately goes into the DB, either because of default values or for some other reasons. This object is validated against the schema already as part of the update call on line 59, so it seemed like checking it here is a bit superfluous anyway.

zenweasel commented on imports/plugins/included/discount-codes/server/methods/methods.app-test.js:
```diff
@@ -17,7 +17,6 @@ const code = {
 
 before(function () {
   this.timeout(10000);
-  Meteor._sleepForMs(7000);
```
We use this sleep to make sure fixtures are completely loaded before running tests on older systems

zenweasel commented on imports/plugins/included/discount-rates/server/methods/methods.app-test.js:
```diff
@@ -42,7 +41,6 @@ describe("discount rate methods", function () {
       expect(discountInsertSpy).to.have.been.called;
 
       const discountCount = Discounts.find(discountId).count();
-      Meteor._sleepForMs(500);
```
I'm getting test failures re-introduced by removing these sleeps.

aldeed replied:
Is it okay if I try to fix the code instead? Needing to sleep for tests is a tenuous fix and just ignores the fact that failures due to timing usually mean poorly written code.

zenweasel replied:
For sure. I would love to get these really fixed. Just pointing out that the tests are failing intermittently now.

zenweasel commented on imports/plugins/included/shipping-shippo/server/methods/shippoapi.js:
```diff
@@ -2,7 +2,7 @@
 import { Meteor } from "meteor/meteor";
 import { ValidatedMethod } from "meteor/mdg:validated-method";
-import { SimpleSchema } from "meteor/aldeed:simple-schema";
+import SimpleSchema from "simpl-schema";
```
import order

zenweasel commented on lib/collections/schemas/products.js:
```diff
@@ -1,8 +1,10 @@
 import { Random } from "meteor/random";
-import { SimpleSchema } from "meteor/aldeed:simple-schema";
+import { Tracker } from "meteor/tracker";
+import SimpleSchema from "simpl-schema";
```
import order

zenweasel commented on server/methods/accounts/accounts.app-test.js:
```diff
@@ -104,6 +104,7 @@ describe("Account Meteor method ", function () {
       const newAddress = account.profile.addressBook[
         account.profile.addressBook.length - 1];
       delete newAddress._id;
+      delete newAddress.failedValidation;
```
Why are we deleting this here?

aldeed replied:
Because it is added to the doc automatically due to a defaultValue. I'm not sure why it wasn't an issue prior to my changes, but I'm pretty sure the defaultValue was always there. If you don't want to delete it, then it could be added to the original address being inserted.

zenweasel replied:
No, I was just trying to understand why we were deleting it

zenweasel commented on server/publications/collections/products.js:
```diff
@@ -1,7 +1,8 @@
+import { Tracker } from "meteor/tracker";
 import { check, Match } from "meteor/check";
-import { SimpleSchema } from "meteor/aldeed:simple-schema";
+import SimpleSchema from "simpl-schema";
```
import order

zenweasel commented on server/startup/init.js:
```diff
@@ -1,6 +1,18 @@
+import { Meteor } from "meteor/meteor";
+import SimpleSchema from "simpl-schema";
```
import order

zenweasel commented:
I don't seem to be getting any validation error when saving settings in any of the settings panels

aaronjudd commented:
This is looking really close, fantastic. 

I see id/react/validation issues in Taxes, Accounts, Login Services but verified that these console errors all exist in master as well, so not apparently introduced here.  The lack of validation, as reported by Brent I am able to replicate as well in the Avalara panel, and is not reproducible in master. Most other forms do seem to have validation though. 

I noticed the sort order of the packages (as seen in the icon-side-bar) are different in master versus this branch. 

Fatal error in checkout address validation, reproducible by adding address in checkout...

```
Exception in delivering result of invoking 'accounts/validateAddress': TypeError: Cannot read property 'validated' of undefined
    at http://localhost:3000/app/app.js?hash=8bb9266b26ebdf2fa9b1a45ac3b8ba21dd556a6c:37967:19
    at MethodInvoker._callback (http://localhost:3000/packages/meteor.js?hash=cbcc712d51de4298c275e8dcf25c66c29914f19a:1167:22)
    at MethodInvoker._maybeInvokeCallback (http://localhost:3000/packages/ddp-client.js?hash=df770fd9a6a02fd730939b97d266ea2b12938e95:3682:12)
    at MethodInvoker.receiveResult (http://localhost:3000/packages/ddp-client.js?hash=df770fd9a6a02fd730939b97d266ea2b12938e95:3702:10)
    at Connection._livedata_result (http://localhost:3000/packages/ddp-client.js?hash=df770fd9a6a02fd730939b97d266ea2b12938e95:4824:9)
    at onMessage (http://localhost:3000/packages/ddp-client.js?hash=df770fd9a6a02fd730939b97d266ea2b12938e95:3528:206)
    at http://localhost:3000/packages/ddp-client.js?hash=df770fd9a6a02fd730939b97d266ea2b12938e95:2908:9
    at Array.forEach (<anonymous>)
    at Function._.each._.forEach (http://localhost:3000/packages/underscore.js?hash=cde485f60699ff9aced3305f70189e39c665183c:149:11)
    at SockJS.self.socket.onmessage (http://localhost:3000/packages/ddp-client.js?hash=df770fd9a6a02fd730939b97d266ea2b12938e95:2907:43)
```
I'll keep poking, but that one at least is a blocker.

zenweasel commented:
@aldeed Don't forget to enter yourself for the [Hack-a-bug-a-thon](getrxn.io/hackabugathon) We are giving away a $1000 gift card and other prizes

aldeed commented:
@zenweasel @aaronjudd So far I have fixed the following:

- Added back `collectionSchema` function with new code and a deprecation warning logged.
- Fixed import order
- Fixed the settings error on startup
- Fixed the error when updating variant price. This was coming not from the initial call to `products/updateProductField` but from the next call triggered by `Products.before.update` hook. The line `const priceRange = ProductRevision.getProductPriceRange(updateId);` was returning `undefined`. I fixed it by updating `getProductPriceRange` to return range objects if it returns early (in this case, if `product.price` is undefined). Let me know if that looks okay. I'm not sure if price being undefined is a symptom of a larger issue.

Still need to fix:
- Lack of validation in some settings forms
- "Fatal error in checkout address validation, reproducible by adding address in checkout..."
- Random test failures / removing test sleeps

aldeed commented:
@zenweasel @aaronjudd This is ready as far as I can tell, if you can verify that all the issues you saw are fixed.

I'm not sure about the timing test failures @zenweasel mentioned. I'm not seeing them, but if you do, let me know which tests exactly and under what circumstances so that I can try to reproduce and fix properly.

zenweasel commented:
@aldeed Awesome! Retesting now

zenweasel commented:
Seeing an error when checking out which I verified is not happening on `master`
### Steps to reproduce
1. Enable Shippo
1. Enable a payment method
1. Attempt to checkout using a non-admin user
1. Enter address
1. When hitting save and continue receive the following error in the server console.



```
  Error updating rates for cart fMcCsi7r7k3TKdfxW { Error: Method Name is required (shipping.$.shipmentQuotes.0.method.name) in Cart update
      at getErrorObject (packages/aldeed:collection2-core/collection2.js:474:17)
      at ns.Collection.doValidate (packages/aldeed:collection2-core/collection2.js:446:13)
      at ns.Collection.Mongo.Collection.(anonymous function) (packages/aldeed:collection2-core/collection2.js:187:25)
      at ns.Collection.Mongo.Collection.(anonymous function) [as update] (packages/dispatch_run-as-user.js:325:19)
      at server/methods/core/shipping.js:221:10
      at Array.forEach (<anonymous>)
      at updateShippingRecordByShop (server/methods/core/shipping.js:208:9)
      at DDPCommon.MethodInvocation.shippingUpdateShipmentQuotes (server/methods/core/shipping.js:299:7)
      at packages/check.js:128:16
      at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
      at Object._failIfArgumentsAreNotAllChecked (packages/check.js:127:41)
      at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1765:18)
      at DDP._CurrentMethodInvocation.withValue (packages/ddp-server/livedata_server.js:1686:15)
      at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
      at resolve (packages/ddp-server/livedata_server.js:1684:36)
      at new Promise (<anonymous>)
      at Server.applyAsync (packages/ddp-server/livedata_server.js:1683:12)
      at Server.apply (packages/ddp-server/livedata_server.js:1622:26)
      at Server.call (packages/ddp-server/livedata_server.js:1604:17)
      at DDPCommon.MethodInvocation.cartSetShipmentAddress (server/methods/core/cart.js:852:12)
      at packages/check.js:128:16
      at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
    invalidKeys:
     [ { name: 'shipping.$.shipmentQuotes.0.method.name',
         type: 'required',
         value: undefined },
       { name: 'shipping.$.shipmentQuotes.0.method.group',
         type: 'required',
         value: undefined },
       { name: 'shipping.$.shipmentQuotes.1.method.name',
         type: 'required',
         value: undefined },
       { name: 'shipping.$.shipmentQuotes.1.method.group',
         type: 'required',
         value: undefined },
       { name: 'shipping.$.shipmentQuotes.2.method.name',
         type: 'required',
         value: undefined },
       { name: 'shipping.$.shipmentQuotes.2.method.group',
         type: 'required',
         value: undefined },
       { name: 'shipping.$.shipmentQuotes.3.method.name',
         type: 'required',
         value: undefined },
       { name: 'shipping.$.shipmentQuotes.3.method.group',
         type: 'required',
         value: undefined },
       { name: 'shipping.$.shipmentQuotes.4.method.name',
         type: 'required',
         value: undefined },
       { name: 'shipping.$.shipmentQuotes.4.method.group',
         type: 'required',
         value: undefined },
       { name: 'shipping.$.shipmentQuotes.5.method.name',
         type: 'required',
         value: undefined },
       { name: 'shipping.$.shipmentQuotes.5.method.group',
         type: 'required',
         value: undefined },
       { name: 'shipping.$.shipmentQuotes.6.method.name',
         type: 'required',
         value: undefined },
       { name: 'shipping.$.shipmentQuotes.6.method.group',
         type: 'required',
         value: undefined } ],
    validationContext:
     ValidationContext {
       _simpleSchema:
        SimpleSchema {
          pick: [Function: pickOrOmit],
          omit: [Function: pickOrOmit],
          _constructorOptions: [Object],
          _validators: [],
          _docValidators: [],
          _validationContexts: [Object],
          _cleanOptions: [Object],
          _schema: [Object],
          _depsLabels: {},
          _schemaKeys: [Array],
          _autoValues: [Array],
          _blackboxKeys: [Array],
          _firstLevelSchemaKeys: [Array],
          _objectKeys: [Object],
          messageBox: [Object],
          version: 2 },
       _schema:
        { _id: [Object],
          shopId: [Object],
          userId: [Object],
          sessionId: [Object],
          email: [Object],
          items: [Object],
          'items.$': [Object],
          shipping: [Object],
          'shipping.$': [Object],
          billing: [Object],
          'billing.$': [Object],
          tax: [Object],
          taxes: [Object],
          'taxes.$': [Object],
          taxRatesByShop: [Object],
          discount: [Object],
          workflow: [Object],
          createdAt: [Object],
          updatedAt: [Object] },
       _schemaKeys:
        [ '_id',
          'shopId',
          'userId',
          'sessionId',
          'email',
          'items',
          'items.$',
          'shipping',
          'shipping.$',
          'billing',
          'billing.$',
          'tax',
          'taxes',
          'taxes.$',
          'taxRatesByShop',
          'discount',
          'workflow',
          'createdAt',
          'updatedAt' ],
       _validationErrors:
        [ [Object],
          [Object],
          [Object],
          [Object],
          [Object],
          [Object],
          [Object],
          [Object],
          [Object],
          [Object],
          [Object],
          [Object],
          [Object],
          [Object] ],
       _deps:
        { _id: [Object],
          shopId: [Object],
          userId: [Object],
          sessionId: [Object],
          email: [Object],
          items: [Object],
          'items.$': [Object],
          shipping: [Object],
          'shipping.$': [Object],
          billing: [Object],
          'billing.$': [Object],
          tax: [Object],
          taxes: [Object],
          'taxes.$': [Object],
          taxRatesByShop: [Object],
          discount: [Object],
          workflow: [Object],
          createdAt: [Object],
          updatedAt: [Object] },
       _depsAny: { _dependentsById: {} } },
    sanitizedError:
     { Error: Method Name is required (shipping.$.shipmentQuotes.0.method.name) in Cart update [400]
      at getErrorObject (packages/aldeed:collection2-core/collection2.js:480:28)
      at ns.Collection.doValidate (packages/aldeed:collection2-core/collection2.js:446:13)
      at ns.Collection.Mongo.Collection.(anonymous function) (packages/aldeed:collection2-core/collection2.js:187:25)
      at ns.Collection.Mongo.Collection.(anonymous function) [as update] (packages/dispatch_run-as-user.js:325:19)
      at server/methods/core/shipping.js:221:10
      at Array.forEach (<anonymous>)
      at updateShippingRecordByShop (server/methods/core/shipping.js:208:9)
      at DDPCommon.MethodInvocation.shippingUpdateShipmentQuotes (server/methods/core/shipping.js:299:7)
      at packages/check.js:128:16
      at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
      at Object._failIfArgumentsAreNotAllChecked (packages/check.js:127:41)
      at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1765:18)
      at DDP._CurrentMethodInvocation.withValue (packages/ddp-server/livedata_server.js:1686:15)
      at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
      at resolve (packages/ddp-server/livedata_server.js:1684:36)
      at new Promise (<anonymous>)
      at Server.applyAsync (packages/ddp-server/livedata_server.js:1683:12)
      at Server.apply (packages/ddp-server/livedata_server.js:1622:26)
      at Server.call (packages/ddp-server/livedata_server.js:1604:17)
      at DDPCommon.MethodInvocation.cartSetShipmentAddress (server/methods/core/cart.js:852:12)
      at packages/check.js:128:16
      at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
       isClientSafe: true,
       error: 400,
       reason: 'Method Name is required (shipping.$.shipmentQuotes.0.method.name) in Cart update',
       details: '[{"name":"shipping.$.shipmentQuotes.0.method.name","type":"required"},{"name":"shipping.$.shipmentQuotes.0.method.group","type":"required"},{"name":"shipping.$.shipmentQuotes.1.method.name","type":"required"},{"name":"shipping.$.shipmentQuotes.1.method.group","type":"required"},{"name":"shipping.$.shipmentQuotes.2.method.name","type":"required"},{"name":"shipping.$.shipmentQuotes.2.method.group","type":"required"},{"name":"shipping.$.shipmentQuotes.3.method.name","type":"required"},{"name":"shipping.$.shipmentQuotes.3.method.group","type":"required"},{"name":"shipping.$.shipmentQuotes.4.method.name","type":"required"},{"name":"shipping.$.shipmentQuotes.4.method.group","type":"required"},{"name":"shipping.$.shipmentQuotes.5.method.name","type":"required"},{"name":"shipping.$.shipmentQuotes.5.method.group","type":"required"},{"name":"shipping.$.shipmentQuotes.6.method.name","type":"required"},{"name":"shipping.$.shipmentQuotes.6.method.group","type":"required"}]',
       message: 'Method Name is required (shipping.$.shipmentQuotes.0.method.name) in Cart update [400]',
       errorType: 'Meteor.Error' } }
```

zenweasel commented:
When I select "Allow Guest Checkout" in the Shop settings I get an error toast:
```
Shop options update failed. Error: Username is required
```

zenweasel commented:
Seeing this error when I add a product in a Marketplace shop and try to add it to Cart:

```
Exception in callback of async function: Error: [Method Name is required]
    at Function.validationErrorTransform (server/startup/init.js:10:20)
    at /Users/brenthoover/Projects/js/reaction/node_modules/simpl-schema/dist/SimpleSchema.js:749:30
    at Array.forEach (<anonymous>)
    at SimpleSchema.validate (/Users/brenthoover/Projects/js/reaction/node_modules/simpl-schema/dist/SimpleSchema.js:722:15)
    at DDPCommon.MethodInvocation.shippingUpdateShipmentQuotes (server/methods/core/shipping.js:291:16)
    at packages/check.js:128:16
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
    at Object._failIfArgumentsAreNotAllChecked (packages/check.js:127:41)
    at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1765:18)
    at DDP._CurrentMethodInvocation.withValue (packages/ddp-server/livedata_server.js:1686:15)
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
    at resolve (packages/ddp-server/livedata_server.js:1684:36)
    at new Promise (<anonymous>)
    at Server.applyAsync (packages/ddp-server/livedata_server.js:1683:12)
    at Server.apply (packages/ddp-server/livedata_server.js:1622:26)
    at Server.call (packages/ddp-server/livedata_server.js:1604:17)
    at server/methods/core/cart.js:474:14
    at wrappedCallbackForParsingMongoValidationErrors (packages/aldeed:collection2-core/collection2.js:506:15)
    at packages/mongo/collection.js:618:5
    at args.(anonymous function) (packages/matb33_collection-hooks.js:500:23)
    at runWithEnvironment (packages/meteor.js:1240:24)
    at packages/meteor.js:1253:14
 => awaited here:
    at Promise.await (/Users/brenthoover/.meteor/packages/promise/.0.10.0.ak8m4i++os+web.browser+web.cordova/npm/node_modules/meteor-promise/promise_server.js:60:12)
    at Server.apply (packages/ddp-server/livedata_server.js:1635:14)
    at Server.call (packages/ddp-server/livedata_server.js:1604:17)
    at server/methods/core/cart.js:474:14
    at wrappedCallbackForParsingMongoValidationErrors (packages/aldeed:collection2-core/collection2.js:506:15)
    at packages/mongo/collection.js:618:5
    at args.(anonymous function) (packages/matb33_collection-hooks.js:500:23)
    at runWithEnvironment (packages/meteor.js:1240:24)
    at packages/meteor.js:1253:14
    at packages/mongo/mongo_driver.js:319:7
    at packages/mongo/mongo_driver.js:606:15
    at runWithEnvironment (packages/meteor.js:1240:24)
```

zenweasel commented:
I get a validation error when I change anything in the "Search" settings panels

aldeed commented:
@zenweasel For the shippo error, I have traced it down to this: https://github.com/reactioncommerce/reaction/blob/master/imports/plugins/included/shipping-shippo/server/methods/shippo.js#L63

When the rate objects are built from the shippo data, the `method` object is not given `name` and `group` properties, but those are required by the schema. They are required in master, too, so the old SimpleSchema must have had a bug that prevented it from catching that schema error.

So, if it is correct for those properties to be required, then what should the `name` be and how do I determine which `group` to use?

zenweasel commented:
@aldeed Just looking at the code and the data I don't have definitive answers. It seems like `name` should be the value in `rate.servicelevel_name` and label should be:
```
    // const methodLabel = `${rate.provider} - ${rate.servicelevel_name}`;
```
as in the commented out code there?

Honestly I don't see how we can ascertain group from the data that Shippo provides.

Since Shippo seems to work without these fields I wonder if we are just safe making them `optional`?

aldeed commented:
@zenweasel I made those optional as we discussed

zenweasel commented:
@aldeed Still seeing a validation error thrown whenever I change any settings in the Search dashboard panel.

zenweasel commented:
Tested this as much as I can and I am only seeing the validation error in Search Settings. Otherwise 👍

aldeed commented:
@zenweasel Should be fixed now, and I merged in latest master

zenweasel commented:
So I was testing how the migration from a previous version would go. I get the following error when attempting to checkout. 

```
in the root directory of your application.
03:29:02.408Z ERROR Reaction:
  An error occurred adding the address { Error: [Public Label is required]
      at Function.validationErrorTransform (server/startup/init.js:10:20)
      at /Users/brenthoover/Projects/js/reaction/node_modules/simpl-schema/dist/SimpleSchema.js:749:30
      at Array.forEach (<anonymous>)
      at SimpleSchema.validate (/Users/brenthoover/Projects/js/reaction/node_modules/simpl-schema/dist/SimpleSchema.js:722:15)
      at DDPCommon.MethodInvocation.discountsCalculate (imports/plugins/core/discounts/server/methods/methods.js:89:27)
      at packages/check.js:128:16
      at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
      at Object._failIfArgumentsAreNotAllChecked (packages/check.js:127:41)
      at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1765:18)
      at DDP._CurrentMethodInvocation.withValue (packages/ddp-server/livedata_server.js:1686:15)
      at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
      at resolve (packages/ddp-server/livedata_server.js:1684:36)
      at new Promise (<anonymous>)
      at Server.applyAsync (packages/ddp-server/livedata_server.js:1683:12)
      at Server.apply (packages/ddp-server/livedata_server.js:1622:26)
      at Server.call (packages/ddp-server/livedata_server.js:1604:17)
      at Object.<anonymous> (imports/plugins/core/discounts/server/hooks/cart.js:22:27)
      at packages/matb33_collection-hooks.js:486:20
      at Array.forEach (<anonymous>)
      at Function._.each._.forEach (packages/underscore.js:139:11)
      at packages/matb33_collection-hooks.js:485:11
      at Array.forEach (<anonymous>)
   => awaited here:
      at Promise.await (/Users/brenthoover/.meteor/packages/promise/.0.10.0.ak8m4i++os+web.browser+web.cordova/npm/node_modules/meteor-promise/promise_server.js:60:12)
      at Server.apply (packages/ddp-server/livedata_server.js:1635:14)
      at Server.call (packages/ddp-server/livedata_server.js:1604:17)
      at Object.<anonymous> (imports/plugins/core/discounts/server/hooks/cart.js:22:27)
      at packages/matb33_collection-hooks.js:486:20
      at Array.forEach (<anonymous>)
      at Function._.each._.forEach (packages/underscore.js:139:11)
      at packages/matb33_collection-hooks.js:485:11
      at Array.forEach (<anonymous>)
      at Function._.each._.forEach (packages/underscore.js:139:11)
      at after (packages/matb33_collection-hooks.js:484:9)
      at Object.<anonymous> (packages/matb33_collection-hooks.js:505:5)
      at Object.collection.(anonymous function) [as update] (packages/matb33_collection-hooks.js:146:21)
      at ns.Collection.update (packages/mongo/collection.js:556:29)
      at ns.Collection.Mongo.Collection.(anonymous function) (packages/aldeed:collection2-core/collection2.js:205:19)
      at ns.Collection.Mongo.Collection.(anonymous function) [as update] (packages/dispatch_run-as-user.js:325:19)
      at server/methods/core/cart.js:785:28
      at Array.forEach (<anonymous>)
      at DDPCommon.MethodInvocation.cartSetShipmentAddress (server/methods/core/cart.js:773:15)
      at packages/check.js:128:16
      at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
      at Object._failIfArgumentsAreNotAllChecked (packages/check.js:127:41)
    isClientSafe: true,
    error: 'validation-error',
    reason: undefined,
    details:
     [ { name: 'shipping.0.shipmentMethod.label',
         type: 'required',
         value: undefined,
         message: 'Public Label is required' },
       { name: 'shipping.0.shipmentMethod.rate',
         type: 'required',
         value: undefined,
         message: 'Rate is required' },
       { name: 'shipping.0.shipmentMethod.enabled',
         type: 'required',
         value: undefined,
         message: 'Enabled is required' } ],
    message: '[Public Label is required]',
    errorType: 'Meteor.Error' }
```

### To Reproduce (all these steps might not be necessary?)
1. Checkout current `master`
1. Configure Flat rate shipping and Authorize.net payment
1. Add a product (I added one with one variant and one option)
1. Checkout
1. Git checkout this branch
1. `rm -rf node_modules`
1. `meteor npm install`
1. Attempt to checkout again
1. Get this above error after entering an address

zenweasel commented:
Also seeing validation errors when saving valid API key for Authorize.net

![2018-01-25_11-48-49](https://user-images.githubusercontent.com/72819/35369952-d2ded892-01c5-11e8-8713-5757bdb65ccf.png)

zenweasel commented:
@aldeed Hey Eric, I tried duplicating this again and still got the same error when following my instructions above. Let me know if you want to screenshare so you can look at my system? Looking at the db all the methods look like they have Label and Rate so I am confused why this error is occurring.

zenweasel commented:
@aaronjudd did you try following the replication steps I had? Maybe adding a product makes a difference?

zenweasel commented:
So working with @aldeed and doing some testing this is my current theory.

When you checkout on the current `master` and the cart is recreated it creates a `shipping[0].shippingMethod` record which I believe is happening because of the `autovalue` here:

https://github.com/reactioncommerce/reaction/blob/f2efb7be3dd7ad9f7af35120c2bba19b2b26b22d/lib/collections/schemas/shipping.js#L61

I think the `blackbox: true` on the `shipping` record was allowing this to work.

So when you switch to this branch and try to check out using the same user and try to modify the cart validation fails because the `shippingMethod` record both exists and is incomplete.

![2018-01-30_13-52-12](https://user-images.githubusercontent.com/72819/35550316-cbf28cd8-05c4-11e8-80c1-a6c70f1da844.png)

Since this isn't really a bug in this branch per-se, maybe the solution is to add a migration to remove all cart shipping records where the `shippingMethod` is incomplete?

aldeed commented:
@zenweasel That sounds right. I added the migration. Pull the latest, make sure the version in Migrations collection is < 19 when you first run my branch, and then see if the error goes away.

zenweasel commented:
Tested and this last issue seems to be fixed.

aldeed commented:
Merge in 1.8.0 and fixed conflicts

DenisBogatirov commented:
Oh my god, this was already working in release branch why you didn't merge it?

aaronjudd commented:
@DenisBogatirov there's still breaking tests and some conflicts that needed cleaning up. Originally we weren't going to release this until the 2.0 release (big changes), but as it's looking pretty close, I've tagged this for 1.9 release (next week), so hopefully it'll be merged pretty quickly.  @aldeed is actually on vacation though, so waiting on him to return, before we do a final review and merge.

pmn4 commented:
the merge conflicts in `shop.js` look pretty straightforward (and conflict with some changes that I made), so happy to lend a hand resolving

aaronjudd commented:
@aldeed new issue found (after merge into 1.9) -> can create a new issue but thought I'd be useful here:


Seen clicking on "Tax" in sidebar, or opening any of the tax settings:


```Exception from Tracker recompute function:
meteor.js?hash=b0f12795c8cc1423b5850502871996903f947ed5:992 TypeError: fieldSchema.type[0].type is not a function
    at Form.renderField (form.js:267)
    at form.js:316
    at Array.map (<anonymous>)
    at Form.renderWithSchema (form.js:304)
    at Form.render (form.js:344)
    at finishClassComponent (modules.js?hash=4f2eb55914217a0342d190203eac0918f9672a78:186673)
    at updateClassComponent (modules.js?hash=4f2eb55914217a0342d190203eac0918f9672a78:186650)
    at beginWork (modules.js?hash=4f2eb55914217a0342d190203eac0918f9672a78:187025)
    at performUnitOfWork (modules.js?hash=4f2eb55914217a0342d190203eac0918f9672a78:189024)
    at workLoop (modules.js?hash=4f2eb55914217a0342d190203eac0918f9672a78:189088)```
